src/enclave/http.h

```diff
@@ -0,0 +1,256 @@
+          state = DONE;
+        }
+        else
+          throw std::runtime_error("Ending message, but not in a message");
+      }
+    };
+
+    class ResponseHeaderEmitter
+    {
+    public:
+      static std::vector<uint8_t> emit(const std::vector<uint8_t> data)
+      {
+        if (data.size() == 0)
+        {
+          auto hdr = fmt::format("HTTP/1.1 204 No Content\r\n");
+          return std::vector<uint8_t>(hdr.begin(), hdr.end());
+        }
+        else
+        {
+          auto hdr = fmt::format(
+            "HTTP/1.1 200 OK\r\nContent-Type: "
+            "application/json\r\nContent-Length: {}\r\n\r\n",
+            data.size());
+          return std::vector<uint8_t>(hdr.begin(), hdr.end());
+        }
+      }
+    };
+
+    class RequestHeaderEmitter
+    {
+    public:
+      static std::vector<uint8_t> emit(const std::vector<uint8_t> data)
+      {
+        return http::post_header(data);
+      }
+    };
+
+    static int on_msg(http_parser* parser)
+    {
+      Parser* p = reinterpret_cast<Parser*>(parser->data);
+      p->new_message();
+      return 0;
+    }
+
+    static int on_msg_end(http_parser* parser)
+    {
+      Parser* p = reinterpret_cast<Parser*>(parser->data);
+      p->end_message();
+      return 0;
+    }
+
+    static int on_req(http_parser* parser, const char* at, size_t length)
+    {
+      Parser* p = reinterpret_cast<Parser*>(parser->data);
+      p->append(at, length);
+      return 0;
+    }
+  }
+
+  template <class E>
+  class HTTPEndpoint : public TLSEndpoint, public http::MsgProcessor
+  {
+  protected:
+    http::Parser p;
+
+  public:
+    HTTPEndpoint(
+      size_t session_id,
+      ringbuffer::AbstractWriterFactory& writer_factory,
+      std::unique_ptr<tls::Context> ctx)
+    {
+      LOG_FAIL_FMT("FAIL");
+      assert(false);
+    }
+
+    void recv(const uint8_t* data, size_t size)
+    {
+      recv_buffered(data, size);
+
+      LOG_TRACE_FMT("recv called with {} bytes", size);
+      auto buf = read(4096, false); // TODO: retry if more was pending
+      LOG_TRACE_FMT("read got {}", buf.size());
+      if (buf.size() == 0)
+        return;
+      LOG_TRACE_FMT(
+        "Going to parse {} bytes: [{}]",
+        buf.size(),
+        std::string(buf.begin(), buf.end()));
+
+      if (p.execute(buf.data(), buf.size()) == 0)
+        return;
+    }
+
+    virtual void msg(std::vector<uint8_t> m)
+    {
+      if (m.size() > 0)
+      {
+        try
+        {
+          if (!handle_data(m))
+            close();
+        }
+        catch (...)
+        {
+          // On any exception, close the connection.
+          close();
+        }
+      }
+    }
+
+    void send(const std::vector<uint8_t>& data)
+    {
+      auto h = E::emit(data);
+      LOG_TRACE_FMT(
+        "Going to send header [{}]", std::string(h.begin(), h.end()));
+      LOG_TRACE_FMT(
+        "Going to send body [{}]", std::string(data.begin(), data.end()));
+
+      send_buffered(E::emit(data));
```

Comment:
`send_buffered(h)`?

Reply:
I forgot to remove the debug above, sorry.